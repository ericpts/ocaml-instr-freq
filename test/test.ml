open Core
open Instr_freq

let find_fixtures_directory () =
  let rec walk cur =
    if cur = Filename.root then
      failwith
        "Could not find fixtures directory! Make sure that you are running \
         the test from the project's root, or a subdirectory.";
    let attempt = cur ^/ "test" ^/ "fixtures" in
    match Sys.is_directory attempt with
    | `Yes -> attempt
    | `No | `Unknown -> walk (Filename.dirname cur)
  in
  walk (Filename.realpath Sys.executable_name)
;;

let matcher_of_sexp sexp =
  let fname = Filename.temp_file "matcher" ".sexp" in
  Out_channel.write_all fname ~data:sexp;
  let matcher =
    let instructions =
      Sexp.load_sexps_conv_exn fname
        [%of_sexp: Index.Matcher.desc Index.With_register_information.t]
    in
    Index.Matcher.create instructions
  in
  Sys.remove fname;
  matcher
;;

let build_index_for_fixture ~fixtures_directory ~context_length fixture_name
    =
  let ml_file = fixtures_directory ^/ sprintf "%s.ml" fixture_name in
  let open Async in
  (* Don't enter the async monad, to preserve error message call stacks, in
     case the tests fail. *)
  Thread_safe.block_on_async_exn (fun () ->
      Process.run_expect_no_output_exn () ~prog:"ocamlopt"
        ~args:
          [ "-save-ir-after";
            "linearize";
            "-stop-after";
            "linearize";
            ml_file
          ]);
  let linear_file = Filename.chop_extension ml_file ^ ".cmir-linear" in
  let blocks =
    Loop_free_block.read_file ~file:linear_file ~context_length
  in
  let index = Index.empty () in
  List.iter blocks ~f:(fun (_fundecl, loop_free_blocks) ->
      Array.iter loop_free_blocks ~f:(fun loop_free_block ->
          Index.update index loop_free_block));
  (blocks, index)
;;

let iter_blocks blocks ~index ~file ~statistics =
  let { Stats.on_block; on_finish_iteration; hint_files = _ } =
    statistics
  in
  List.iter blocks ~f:(fun (fundecl, loop_free_blocks) ->
      Array.iter loop_free_blocks ~f:(fun block ->
          let equivalence = Index.equivalence_exn index block in
          let frequency = Index.frequency_exn index equivalence in
          on_block block ~file ~equivalence ~frequency ~fundecl |> ignore));
  on_finish_iteration ()
;;

let test_index_statistics ~fixtures_directory =
  let _blocks, index =
    build_index_for_fixture "simple" ~fixtures_directory ~context_length:0
  in
  let { Index.n_symbolic_blocks;
        n_basic_instructions;
        n_terminator_instructions
      } =
    Index.hashtbl_load_statistics index
  in
  [%test_result: int] ~expect:8 n_symbolic_blocks;
  [%test_result: int] ~expect:11 n_basic_instructions;
  [%test_result: int] ~expect:4 n_terminator_instructions
;;

let test_matching_functionality ~fixtures_directory =
  printf "test_matching_functionality\n";
  let blocks, index =
    build_index_for_fixture "simple" ~fixtures_directory ~context_length:0
  in
  let matcher =
    let sexp = {|
(
 (desc (Basic (Op Move)))
 (arg (0))
 (res (1))
)
|} in
    Some (matcher_of_sexp sexp index)
  in
  let statistics =
    Stats.combine
      [ Stats.count_blocks_matching index ~min_block_size:0 ~matcher;
        Stats.print_most_popular_classes index ~n_real_blocks_to_print:1
          ~n_most_frequent_equivalences:1
          ~block_print_mode:Loop_free_block.As_assembly ~min_block_size:0
          ~matcher
      ]
  in
  iter_blocks blocks ~index ~file:"simple.ml" ~statistics
;;

let test_index_bigger_file ~fixtures_directory =
  printf "test_index_bigger_file\n";
  let blocks, index =
    build_index_for_fixture "repetitive" ~fixtures_directory
      ~context_length:0
  in
  let statistics =
    Stats.combine
      [ Stats.print_most_popular_classes index ~n_real_blocks_to_print:1
          ~n_most_frequent_equivalences:1
          ~block_print_mode:Loop_free_block.As_cfg ~min_block_size:4
          ~matcher:None
      ]
  in
  iter_blocks blocks ~index ~file:"repetitive.ml" ~statistics
;;

let test_index_context ~fixtures_directory =
  printf "test_index_context\n";
  let blocks, index =
    build_index_for_fixture "repetitive" ~fixtures_directory
      ~context_length:1
  in
  let statistics =
    Stats.combine
      [ Stats.print_most_popular_classes index ~n_real_blocks_to_print:1
          ~n_most_frequent_equivalences:1
          ~block_print_mode:Loop_free_block.Both ~min_block_size:6
          ~matcher:None
      ]
  in
  iter_blocks blocks ~index ~file:"repetitive.ml" ~statistics
;;

let test_index_context_with_matcher ~fixtures_directory =
  printf "test_index_context_with_matcher\n";
  let blocks, index =
    build_index_for_fixture "long_function" ~fixtures_directory
      ~context_length:2
  in
  let matcher =
    let sexp =
      {|
(
 (desc (Basic (Op (Intop_imm Iadd -2))))
 (arg (0))
 (res (0))
)
(
 (desc (Terminator
        (Branch
         (((Test (Iinttest (Isigned Cgt))) 101)
          ((Test (Iinttest (Isigned Cle))) 127)))))
 (arg (1 0))
 (res ())
)
(
 (desc (Basic (Op Spill)))
 (arg (0))
 (res (2))
)
|}
    in
    Some (matcher_of_sexp sexp index)
  in
  let statistics =
    Stats.combine
      [ Stats.print_most_popular_classes index ~n_real_blocks_to_print:1
          ~n_most_frequent_equivalences:1
          ~block_print_mode:Loop_free_block.Both ~min_block_size:0 ~matcher
      ]
  in
  iter_blocks blocks ~index ~file:"repetitive.ml" ~statistics
;;

let main ~fixtures_directory =
  test_index_statistics ~fixtures_directory;
  test_matching_functionality ~fixtures_directory;
  test_index_bigger_file ~fixtures_directory;
  test_index_context ~fixtures_directory;
  test_index_context_with_matcher ~fixtures_directory
;;

let () = main ~fixtures_directory:(find_fixtures_directory ())
