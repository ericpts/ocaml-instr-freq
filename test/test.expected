test_matching_functionality
There are a total of 2 blocks matching the query criteria.
Equivalence [38;5;6m<6>[0m with 1 members;
	file: [38;5;6msimple.ml[0m
	function: [38;5;6mcamlSimple__impl_36[0m
[38;5;2mBlocks (105): {[0m
	.file ""
	.section .text._fun_start_,"ax",@progbits
	.align	16
	.globl	_fun_start_
_fun_start_:
	.cfi_startproc
.L105:
	movq	%rax, %rdi
	sarq	$1, %rdi
	decq	%rbx
	imulq	%rdi, %rbx
	incq	%rbx
	addq	$-2, %rax
	jmp	camlSimple__impl_36@PLT
	.cfi_endproc
	.type _fun_start_,@function
	.size _fun_start_,. - _fun_start_
[38;5;2m}[0m
test_index_bigger_file
Equivalence [38;5;6m<1>[0m with 3 members;
	file: [38;5;6mrepetitive.ml[0m
	function: [38;5;6mcamlRepetitive__f0_11[0m
[38;5;2mBlocks (109): {[0m
.L109:
	(Op Spill): Arg(reg#0) Res(stack#local#0) Live(reg#0)
	(Op (Intop_imm Iadd -4)): Arg(reg#0) Res(reg#0) Live(stack#local#0)
	(Call (F (Immediate (func camlRepetitive__f0_11) (label_after 100)))): Arg(reg#0) Res(reg#0) Live(stack#local#0)
	(Op Spill): Arg(reg#0) Res(stack#local#1) Live(stack#local#0)
	(Op Reload): Arg(stack#local#0) Res(reg#0) Live(stack#local#1)
	(Op (Intop_imm Iadd -2)): Arg(reg#0) Res(reg#0) Live(stack#local#1)
	(Call (F (Immediate (func camlRepetitive__f0_11) (label_after 101)))): Arg(reg#0) Res(reg#0) Live(stack#local#1)
	(Op Reload): Arg(stack#local#1) Res(reg#1) Live(reg#0)
	(Op (Intop Iadd)): Arg(reg#0 reg#1) Res(reg#0) Live()
	(Op (Intop_imm Iadd -1)): Arg(reg#0) Res(reg#0) Live()
	Reloadretaddr: Arg() Res() Live()
	#-Return-# Arg(reg#0) Res() Live() 
[38;5;2m}[0m
test_index_context
Equivalence [38;5;6m<0>[0m with 3 members;
	file: [38;5;6mrepetitive.ml[0m
	function: [38;5;6mcamlRepetitive__f0_11[0m
[38;5;2mBlocks (0 109): {[0m
	.file ""
	.section .text._fun_start_,"ax",@progbits
	.align	16
	.globl	_fun_start_
_fun_start_:
	.cfi_startproc
.L0:
.L112:
	cmpq	$3, %rax
	jbe	.L102
.L109:
	movq	%rax, (%rsp)
	addq	$-4, %rax
	call	camlRepetitive__f0_11@PLT
.L100:
	movq	%rax, 8(%rsp)
	movq	(%rsp), %rax
	addq	$-2, %rax
	call	camlRepetitive__f0_11@PLT
.L101:
	movq	8(%rsp), %rbx
	addq	%rbx, %rax
	decq	%rax
	ret
	.cfi_endproc
	.type _fun_start_,@function
	.size _fun_start_,. - _fun_start_
[38;5;2m==========[0m
.L0:
	Prologue: Arg() Res() Live()
	#-(Branch
	    (((Test (Iinttest_imm (Iunsigned Cle) 3)) 102)
	        ((Test (Iinttest_imm (Iunsigned Cgt) 3)) 109)))-# Arg(reg#0) Res() Live() 
.L109:
	(Op Spill): Arg(reg#0) Res(stack#local#0) Live(reg#0)
	(Op (Intop_imm Iadd -4)): Arg(reg#0) Res(reg#0) Live(stack#local#0)
	(Call (F (Immediate (func camlRepetitive__f0_11) (label_after 100)))): Arg(reg#0) Res(reg#0) Live(stack#local#0)
	(Op Spill): Arg(reg#0) Res(stack#local#1) Live(stack#local#0)
	(Op Reload): Arg(stack#local#0) Res(reg#0) Live(stack#local#1)
	(Op (Intop_imm Iadd -2)): Arg(reg#0) Res(reg#0) Live(stack#local#1)
	(Call (F (Immediate (func camlRepetitive__f0_11) (label_after 101)))): Arg(reg#0) Res(reg#0) Live(stack#local#1)
	(Op Reload): Arg(stack#local#1) Res(reg#1) Live(reg#0)
	(Op (Intop Iadd)): Arg(reg#0 reg#1) Res(reg#0) Live()
	(Op (Intop_imm Iadd -1)): Arg(reg#0) Res(reg#0) Live()
	Reloadretaddr: Arg() Res() Live()
	#-Return-# Arg(reg#0) Res() Live() 
[38;5;2m}[0m
test_index_context_with_matcher
Equivalence [38;5;6m<3>[0m with 2 members;
	file: [38;5;6mrepetitive.ml[0m
	function: [38;5;6mcamlLong_function__matrix_multiply_8[0m
[38;5;2mBlocks (118 107 119): {[0m
	.file ""
	.section .text._fun_start_,"ax",@progbits
	.align	16
	.globl	_fun_start_
_fun_start_:
	.cfi_startproc
.L118:
	movq	$1, %rbx
	movq	%rbx, 8(%rsp)
.L107:
	movq	$1, %rdi
	call	camlStdlib__array__make_matrix_142@PLT
.L100:
	movq	$1, %rdx
	movq	(%rsp), %rbx
	addq	$-2, %rbx
	cmpq	%rbx, %rdx
	jg	.L101
.L119:
	movq	%rbx, (%rsp)
	jmp	.L102
	.cfi_endproc
	.type _fun_start_,@function
	.size _fun_start_,. - _fun_start_
[38;5;2m==========[0m
.L118:
	(Op (Const_int 1)): Arg() Res(reg#1) Live(stack#local#0 stack#local#2 stack#local#3 stack#local#4 reg#0)
	(Op Spill): Arg(reg#1) Res(stack#local#1) Live(stack#local#0 stack#local#2 stack#local#3 stack#local#4 reg#1 reg#0)
	#-(Branch ((Always 107)))-# Arg() Res() Live() 
.L107:
	(Op (Const_int 1)): Arg() Res(reg#2) Live(stack#local#0 stack#local#1 stack#local#2 stack#local#3 stack#local#4 reg#1
	    reg#0)
	(Call
	    (F
	        (Immediate (func camlStdlib__array__make_matrix_142)
	            (label_after 100)))): Arg(reg#0 reg#1 reg#2) Res(reg#0) Live(stack#local#0 stack#local#1 stack#local#2 stack#local#3 stack#local#4)
	(Op (Const_int 1)): Arg() Res(reg#4) Live(stack#local#0 stack#local#1 stack#local#2 stack#local#3 stack#local#4 reg#0)
	(Op Reload): Arg(stack#local#0) Res(reg#1) Live(stack#local#1 stack#local#2 stack#local#3 stack#local#4 reg#4 reg#0)
	(Op (Intop_imm Iadd -2)): Arg(reg#1) Res(reg#1) Live(stack#local#1 stack#local#2 stack#local#3 stack#local#4 reg#4 reg#0)
	#-(Branch
	    (((Test (Iinttest (Isigned Cgt))) 101)
	        ((Test (Iinttest (Isigned Cle))) 119)))-# Arg(reg#4 reg#1) Res() Live() 
.L119:
	(Op Spill): Arg(reg#1) Res(stack#local#0) Live(stack#local#1 stack#local#2 stack#local#3 stack#local#4 reg#4 reg#0)
	#-(Branch ((Always 102)))-# Arg() Res() Live() 
[38;5;2m}[0m
